src/Microsoft.AspNet.OData.Shared/Query/AutoSelectExpandHelper.cs

```diff
@@ -347,5 +337,103 @@ private static bool CanExpand(this IEdmModel edmModel, IEdmStructuredType struct
                 }
             }
         }
+
+        /// <summary>
+        /// This is a helper that allows us to avoid inefficiencies in the previous pattern:
+        ///     var structuredTypes = new List&lt;IEdmStructuredType&gt;();
+        ///     structuredTypes.Add(structuredType);
+        ///     structuredTypes.AddRange(edmModel.FindAllDerivedTypes(structuredType));
+        ///
+        /// Specifically, the allocation of the list and the resizing driven by the "AddRange" call are
+        /// avoided by leveraging a struct with a simple state machine for enumerating over a type
+        /// and its derived types.
+        /// </summary>
+        private struct SelfAndDerivedEnumerator : IEnumerator<IEdmStructuredType>
+        {
+            private enum Stage : byte
+            {
+                Initial,
+                Self,
+                Derived,
+            }
+
+            private readonly IEnumerator<IEdmStructuredType> derivedEnumerator;
+            private readonly IEdmStructuredType structuredType;
+
+            private Stage stage;
+
+            public SelfAndDerivedEnumerator(IEdmStructuredType structuredType, IEdmModel edmModel)
+            {
+                if (structuredType == null)
```

Comment:
Given that this is all internal, these `null` checks probably make more sense as `Debug.Assert`s